File: src/my_robot_tutorials/scripts/battery_node.py
#!/usr/bin/env python3

import rospy
from my_robot_msgs.srv import SetLed


def set_led_routine(on):
    try:
        # Create a function that can send requests to the service desired
        set_led = rospy.ServiceProxy(
            "/set_led",  # Service Name
            SetLed  # Service (or request type)
        )
        response = set_led(3, on)
        rospy.loginfo("LED 3 set to: " + str(on == 1))
    except rospy.ServiceException as e:
        rospy.logwarn("Service failed: " + str(e))


if __name__ == "__main__":
    rospy.init_node("battery_node")

    # The code will wait here until the desired service is created
    rospy.wait_for_service("/set_led")

    while not rospy.is_shutdown():
        rospy.sleep(7)
        set_led_routine(1)
        rospy.sleep(3)
        set_led_routine(0)
          

File: src/my_robot_tutorials/src/battery_node.cpp
#include <ros/ros.h>
#include <my_robot_msgs/SetLed.h>



// Define the service and its request
my_robot_msgs::SetLed srv;
    

void set_led_routine(ros::ServiceClient client, bool on){
    srv.request.led_number = 3;
    srv.request.state = on ? 1 : 0;
    if (client.call(srv)) {  //If service call is succesful
        ROS_INFO("Battery %s. Turning LED 3 %s", on ? "empty" : "full", on ? "ON" : "OFF");
    }
    else {
        ROS_WARN("Service Call failed");
    }
}


int main(int argc, char **argv)
{
    ros::init(argc, argv, "battery_node");
    ros::NodeHandle nh;
    ROS_INFO("Battery Node started");

    //                                             type or request             name of service 
    ros::ServiceClient client = nh.serviceClient<my_robot_msgs::SetLed>("/set_led");


    // Wait until there is a server that serves the desired service
    // -1: means wait indefinitely
    // 1 means wait for 1 second
    // 2 means wait for 2 seconds, etc
    ros::service::waitForService("/set_led", ros::Duration(-1));

    while(ros::ok()) {
        // Generate random number
        ros::Duration(3.0).sleep();
        set_led_routine(client, false);   
        ros::Duration(7.0).sleep();
        set_led_routine(client, true);
    }
}